File: 14-prime-cryptarithm/crypt1.cpp
/*
ID: caleb.h1
LANG: C++
PROG: crypt1
*/

#include <fstream>
#include <unordered_set>

#include <iostream>

bool hasBadDigits(uint n, const std::unordered_set<char>& digits) {
    while (n) {
        // If there's a bad digit, return true
        if (digits.find(n % 10) == digits.end()) {
            return true;
        }
        // shift n down a digit
        n /= 10;
    }
    return false;
}

int main() {
    std::unordered_set<char> digits;
    uint solutionsCount = 0;
    std::ifstream fin("crypt1.in");
    int N, t;
    fin >> N;
    for (char i = 0; i < N; ++i) {
        fin >> t;
        digits.insert(t);
    }

    // firstNum is 3 digits
    for (uint firstNum = 100; firstNum < 1000; ++firstNum) {
        // firstNum has viable digits
        if (hasBadDigits(firstNum, digits))
            continue;
        // secondNum is 2 digits
        for (uint secondNum = 10; secondNum < 100; ++secondNum) {
            // seconNum has viable digits
            if (hasBadDigits(secondNum, digits))
                continue;

            uint firstProduct = firstNum * (secondNum % 10);

            // firstProduct is 3 digits
            if (firstProduct < 100 || firstProduct > 999)
                continue;
            // firstProduct contains viable numbers
            if (hasBadDigits(firstProduct, digits))
                continue;

            uint secondProduct = firstNum * (secondNum / 10);

            // secondProduct is 3 digits
            if (secondProduct < 100 || secondProduct > 999)
                continue;
            // secondProduct has viable numbers
            if (hasBadDigits(secondProduct, digits))
                continue;
            
            uint product = firstProduct + secondProduct * 10;
            
            // product has 4 digits
            if (product < 1000 || product > 9999)
                continue;
            if (hasBadDigits(product, digits))
                continue;
            
            ++solutionsCount;
        }
    }

    std::ofstream fout("crypt1.out");
    fout << solutionsCount << std::endl;
    fout.close();

    return 0;
}

File: 14-prime-cryptarithm/crypt1.py
'''
ID: caleb.h1
LANG: PYTHON3
PROG: crypt1
'''

def hasBadDigits(n, digits):
    for char in str(n):
        if not char in digits:
            return True
    return False


def main():
    digits = set()
    solutionsCount = 0
    with open("crypt1.in", "r") as f:
        f.readline()

        for i in f.readline().split():
            digits.add(i)
    
    # firstNum is 3 digits
    for firstNum in range(100, 1000):
        # firstNum has viable digits
        if hasBadDigits(firstNum, digits):
            continue
        # secondNum is 2 digits
        for secondNum in range(10, 100):
            # seconNum has viable digits
            if hasBadDigits(secondNum, digits):
                continue

            firstProduct = firstNum * (secondNum % 10)


            # firstProduct is 3 digits
            if firstProduct < 100 or firstProduct > 999:
                continue
            # firstProduct contains viable numbers
            if hasBadDigits(firstProduct, digits):
                continue

            secondProduct = firstNum * (secondNum // 10)


            # secondProduct is 3 digits
            if secondProduct < 100 or secondProduct > 999:
                continue
            # secondProduct has viable numbers
            if hasBadDigits(secondProduct, digits):
                continue
            
            product = firstProduct + secondProduct * 10
            
            # product has 4 digits
            if product < 1000 or product > 9999:
                continue
            if hasBadDigits(product, digits):
                continue

            solutionsCount += 1

    with open("crypt1.out", "w") as f:
        f.write(f'{solutionsCount}\n')

if __name__ == '__main__':
    main()